### DISCUSSION TITLE
[SvelteKitAuth & Keycloak] OperationProcessingError: "response" is not a conform Authorization Server Metadata response

### DISCUSSION THREAD

**Discussion by rhumbertgz** [nextauthjs/next-auth#9214] [Help]
Hello,

I modified the [https://github.com/nextauthjs/next-auth/tree/main/apps/examples/sveltekit ](https://github.com/nextauthjs/next-auth/tree/main/apps/examples/sveltekit ) demo to use Keycloak as provider instead Github, but whenever I try to login I get the following error: 

```
[auth][error] OperationProcessingError: "response" is not a conform Authorization Server Metadata response
    at Module.processDiscoveryResponse (file:///Users/humberto/Downloads/next-auth-main/apps/examples/sveltekit/node_modules/oauth4webapi/build/index.js:215:15)
    at getAuthorizationUrl (file:///Users/humberto/Downloads/next-auth-main/apps/examples/sveltekit/node_modules/@auth/core/lib/actions/signin/authorization-url.js:19:28)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Module.signIn (file:///Users/humberto/Downloads/next-auth-main/apps/examples/sveltekit/node_modules/@auth/core/lib/actions/signin/index.js:10:56)
    at async AuthInternal (file:///Users/humberto/Downloads/next-auth-main/apps/examples/sveltekit/node_modules/@auth/core/lib/index.js:58:24)
    at async Proxy.Auth (file:///Users/humberto/Downloads/next-auth-main/apps/examples/sveltekit/node_modules/@auth/core/index.js:104:29)
    at async Module.respond (/node_modules/@sveltejs/kit/src/runtime/server/respond.js:282:20)
    at async file:///Users/humberto/Downloads/next-auth-main/apps/examples/sveltekit/node_modules/@sveltejs/kit/src/exports/vite/dev/index.js:510:22
```

![image](https://github.com/nextauthjs/next-auth/assets/10393245/f5d07b15-87e8-45e6-9e7c-1712b1fa4789)

Here is my updated `hooks.server.ts` file:

```
import { SvelteKitAuth } from "@auth/sveltekit"

import Keycloak from "@auth/core/providers/keycloak"
import { KEYCLOAK_URL, KEYCLOAK_REALM, KEYCLOAK_CLIENT_ID, KEYCLOAK_CLIENT_SECRET } from "$env/static/private"

const KEYCLOAK_ISSUER = `http://${KEYCLOAK_URL}/realms/${KEYCLOAK_REALM}`
// e.g., http://keycloak.local/realms/svelte

export const handle = SvelteKitAuth({
  providers: [Keycloak({ clientId: KEYCLOAK_CLIENT_ID, clientSecret: KEYCLOAK_CLIENT_SECRET, issuer: KEYCLOAK_ISSUER })],
})

```

My `.env` file:

```
KEYCLOAK_URL=keycloak.local
KEYCLOAK_REALM=svelte
KEYCLOAK_CLIENT_ID=demo
KEYCLOAK_CLIENT_SECRET=MiKsJyuLF8ZntC18uIqqMBZ529veebTr
AUTH_SECRET=5ea1a6bd5fad1eb1d9c4c5a79893d3f8
```

Any idea what is root of the problem?

Thanks for any help in advanced!

//////// ///////// ///////// /////// /////////// //////// //////// //////

I'm running a local instance of Keycloak where I created a realm: svelte

![image](https://github.com/nextauthjs/next-auth/assets/10393245/a8c42905-cc38-43e1-a755-4caffead2228)

 with a client: `demo` 

![image](https://github.com/nextauthjs/next-auth/assets/10393245/953e9df5-0a91-47ff-bc01-fd369980993c)

a single user: `slvelte` that is a part of the `dev` group

![image](https://github.com/nextauthjs/next-auth/assets/10393245/57d7928e-4a28-4233-bb9f-e1a0a3e5ebfd)

**Comment by aparcanapavel** [comment]
Hi @rhumbertgz 

I recently encountered this issue. First things first, the way you have the `KEYCLOAK_ISSUER` saved me a lot of time, because I had `/auth/realms/ropah` and kept getting the error, so thank you!

If you are still running into this, I was able to solve the issue by using the secondary configuration mentioned in the [docs](https://authjs.dev/reference/sveltekit#usage). Similar to you, I am using Keycloak, but I got the configuration from their `@auth/sveltekit/providers/keycloak` repository, not the core. My config looks like this:

```ts
// hooks.server.ts
import { SvelteKitAuth } from "@auth/sveltekit"
import Keycloak from "@auth/sveltekit/providers/keycloak"
import { KEYCLOAK_CLIENT_ID, KEYCLOAK_CLIENT_SECRET, KEYCLOAK_ISSUER, AUTH_SECRET } from "$env/static/private"
import type { Handle } from "@sveltejs/kit";

export const handle = SvelteKitAuth(async (event) => {
  const authOptions = {
    providers: [
      Keycloak({ 
        clientId: KEYCLOAK_CLIENT_ID, 
        clientSecret: KEYCLOAK_CLIENT_SECRET,
        issuer: KEYCLOAK_ISSUER
      })],
    secret: AUTH_SECRET,
    trustHost: true
  }
  return authOptions
}) satisfies Handle;
```

For the `secret: AUTH_SECRET` value, you will need to generate a secret. As per the docs, if you are on UNIX systems you can use `openssl rand -hex 32` or check out `https://generate-secret.vercel.app/32` and add it to your `.env.local` file.

You may also want to setup your user fully, without the email verification or temporary password. Your user should also belong to a group, and that group should have a role tied to it. Here is an article that goes through the steps ([Medium](https://medium.com/theoptimaltechnologist/what-is-keycloak-how-to-use-it-an-example-with-nodejs-part-1-25434e963fed), 2020)

Here is my `.env` file
```bash
KEYCLOAK_CLIENT_ID=<my-client>
KEYCLOAK_CLIENT_SECRET=<super-secret-string>
KEYCLOAK_ISSUER=http://keycloak.local:8080/realms/<my-client>
AUTH_SECRET=<generated-super-secret-string>
```

My application is running on `http://myclient.local:5173`
